### PR TITLE
support basic training (linear layer) for custom data type 

### PR DESCRIPTION
**Pull request by serach24**

### CODE REVIEW

File: tensorflow/core/platform/cus.cc [hunk 1]
```diff
@@ -15,45 +15,23 @@ limitations under the License.
...
+ForceCompile(CastF32ToValue);
+ForceCompile(CastValueToF32);
+ForceCompile(CusAdd);
+ForceCompile(CusSub);
+ForceCompile(CusMul);
+ForceCompile(CusDiv);
+ForceCompile(CusNeg);
+ForceCompile(CusEq);
+ForceCompile(CusNe);
+ForceCompile(CusLt);
```
Comment (serach24) [inline tensorflow/core/platform/cus.cc:34]: Not sure if there is any more elegant solution to make clang compile those unused functions in the cus.h

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc [hunk 2]
```diff
@@ -454,75 +454,47 @@ StatusOr<llvm::Value*> ElementalIrEmitter::EmitCusUnaryOp(
           primitive_util::BitWidth(from_type),
           primitive_util::BitWidth(to_type));
     }
-    // case HloOpcode::kExp:
-    //   return EmitExp(op->shape().element_type(), operand_value);
+    case HloOpcode::kExp:{
+      // todo(chenhao) cast to f32 to calculate for now
+      operand_value = EmitCusToF32(operand_value, b_);
+      auto value = EmitExp(op->shape().element_type(), operand_value);
+      return EmitF32ToCus(value.ValueOrDie(), b_);
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Instead of emitting the casting operation here, can you do the casting inside the exp function defined under struct cus?

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc [hunk 3]
```diff
@@ -454,75 +454,47 @@ StatusOr<llvm::Value*> ElementalIrEmitter::EmitCusUnaryOp(
...
+    }
     // case HloOpcode::kExpm1:
-    //   return EmitExpm1(op->shape().element_type(), operand_value);
-    // case HloOpcode::kLog:
-    //   return EmitLog(op->shape().element_type(), operand_value);
+    case HloOpcode::kLog:{
+      // todo(chenhao) cast to f32 to calculate for now
+      operand_value = EmitCusToF32(operand_value, b_);
+      auto value = EmitLog(op->shape().element_type(), operand_value);
+      return EmitF32ToCus(value.ValueOrDie(), b_);
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: same goes for here and all the other elementary ops down below.

File: tensorflow/compiler/xla/service/gpu/gemm_thunk.cc [hunk 4]
```diff
@@ -124,6 +130,7 @@ static bool DoGemmWithAlgorithm(
                                             : se::blas::Transpose::kNoTranspose;
   auto k = lhs_matrix.transpose ? lhs_matrix.num_rows : lhs_matrix.num_cols;
 
+
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/gpu/gemm_thunk.cc]: remove

File: tensorflow/compiler/xla/service/gpu/gemm_thunk.cc [hunk 5]
```diff
@@ -140,8 +147,7 @@ static bool DoGemmWithAlgorithm(
             *algorithm, output_profile_result)
         .ok();
   }
-
-  if (batch_size != 1) {
+  else if (batch_size != 1) {
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/gpu/gemm_thunk.cc]: "}" and "else if () {" on the same line

Comment (anandj91) [reply tensorflow/compiler/xla/service/gpu/gemm_thunk.cc]: why did you change this line?

File: tensorflow/compiler/xla/service/gpu/gemm_thunk.cc [hunk 6]
```diff
@@ -288,6 +293,12 @@ Status RunGemm(const GpuGemmConfig &gemm_config,
             batch_size, lhs_matrix, rhs_matrix, output_matrix, alpha.real(),
             beta, stream, best_algorithm,
             /*output_profile_result=*/profile_result);
+      case CUS:
+        CHECK_EQ(alpha.imag(), 0);
+        return DoGemmWithAlgorithm<cus, double>(
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/gpu/gemm_thunk.cc:297]: why is it cus and double? What are these types used for?

File: tensorflow/core/framework/types.h [hunk 7]
```diff
@@ -481,12 +481,16 @@ inline bool DataTypeCanUseMemcpy(DataType dt) {
 }
 
 // Returns true iff 'dt' is a real, non-quantized floating point type.
-constexpr DataTypeSet kDataTypeIsFloating =
-    ToSet(DT_HALF) | ToSet(DT_BFLOAT16) | ToSet(DT_CUS) | ToSet(DT_FLOAT) | ToSet(DT_DOUBLE);
+constexpr DataTypeSet kDataTypeIsFloating = ToSet(DT_HALF) |
+                                            ToSet(DT_BFLOAT16) | ToSet(DT_CUS) |
+                                            ToSet(DT_FLOAT) | ToSet(DT_DOUBLE);
```
Comment (anandj91) [inline tensorflow/core/framework/types.h]: We shouldn't set cus as floating point.

File: tensorflow/core/kernels/BUILD [hunk 8]
```diff
@@ -1,5 +1,10 @@
 load("//tensorflow/core/platform:rules_cc.bzl", "cc_library")
 load("@bazel_skylib//rules:build_test.bzl", "build_test")
+load("@local_config_cuda//cuda:build_defs.bzl", "cuda_library")
+load(
+    "@local_config_nccl//:build_defs.bzl",
```
Comment (anandj91) [inline tensorflow/core/kernels/BUILD]: why do you need nccl?

File: tensorflow/core/ops/math_ops.cc [hunk 9]
```diff
@@ -206,19 +206,18 @@ REGISTER_OP("ComplexAbs")
       .Output("y: T")                                                      \
       .Attr(                                                               \
           "T: {bfloat16, cus, half, float, double, int8, int16, int32, int64, " \
-          "complex64, complex128}")                                        \
-      .SetShapeFn(shape_inference::UnchangedShape)
```
Comment (anandj91) [inline tensorflow/core/ops/math_ops.cc:210]: Why did you remove this line?

File: tensorflow/core/platform/BUILD [hunk 10]
```diff
@@ -146,7 +146,6 @@ cc_library(
 cc_library(
     name = "cus",
     hdrs = ["cus.h"],
-    srcs = ["cus.cc"],
```
Comment (anandj91) [inline tensorflow/core/platform/BUILD:149]: do we only have cus.h now?

File: tensorflow/core/platform/cus.h [hunk 11]
```diff
@@ -64,17 +64,17 @@ struct cus {
...
+inline CUSTOM_DEVICE_FUNC cus CusSub(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC cus CusMul(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC cus CusDiv(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC cus CusNeg(const cus a);
+inline CUSTOM_DEVICE_FUNC bool CusEq(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC bool CusNe(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC bool CusLt(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC bool CusLe(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC bool CusGt(const cus a, const cus b);
+inline CUSTOM_DEVICE_FUNC bool CusGe(const cus a, const cus b);
```
Comment (anandj91) [inline tensorflow/core/platform/cus.h]: I think clang is removing these functions because you defined them as inline

File: third_party/gpus/find_cuda_config.py.gz.base64 [hunk 12]
```diff
@@ -1 +1 @@
...
...69ipmPJQcvgsNrjc6GWdWwfPSFiaOcFxrt9P0mu6fqTxOhE641WkgGVXnrXD01Z//H70/R+qs1i77JEOB80zsYzSRR6v5J4D2llWv3iqw9SlP+yMDtGDF0jxWhtooCRFXe325D02j5G80gvvyEHH4Ix+1zoiv2686gef/+of3P71gA33um+JsbvHtxVXB//5r2A4TaO79kD0rb2b1Dj2uEltu9UZM/CDcfR3acjV9G4oQtlE7mXvybvktJRyyLAf5N1e6lYNpdHKZWhLywSPqy46bWd+FJQCmcc8FL53d0cHMk0FxwaKh9rm9Lomj6OQbgfK6Hq7dZeUKXWRLFQXIh2zW55NlwHM9eDu5XFvfEFbjFd5gL9vj/rDYae2zdo5IOs59wchZ8SYdWLPGT/Puq9q44TyrJPFtT5pS7R1lsX4DKOlR1I8AW25HEdVnFGHuf/xqxxX2glHh49LCRonj9t4rF2H5FmH9ZQfpDPYi/kYXKtfP5dQdeAUY1S5wRkbie7W3i05VC1p14wmtiJEGyQdEPLdyxxMTMZs2mV/Yv7RYY8dlVHfG/ZjbQ+V2r30yHN1W4vcuVSed8Q9TMZWXH3rh0NrJQSTkw0aad5OVeqkK/Aqzm/YtWhS4W59ahsNtRnhi0oZ5HhYSrnWRU04tW0fjeKpNva8TED2timbmvrGkDYx0VLu1xSS1UFNRM4yeaOA9IaHl4nH3EZiUuEuorcJBtfHvqZcTPw1sdiLjo1SUSvHa4XSsNZudOysQ7Yakix1fVVDsruoG5JbVGw2pLJQ/EJDssrwNjX1smOTmNTElCQtM1OSVFOTGzB79bi2YXZyV9iap/7qSr+NKmGs3RoEugttBne03tHMn1wKWcfhAKaAm808ums3jVxKXM/l01pys0jtuR1W3FJlroFVqpa181nLRhqYdKqJDRyatyxuYM+sIFe0OSXFkiudsDZxViaz7dw1JlYNHDaUAhq4dC+C3MCpW2W06W2oDBDX6Iw+9/RdWIQ1KsSsTAlAKp+BsrG8z8ujDJ6uLbLeYheeUVKjwO7zEIXk5DhLsyChrrKTmcSMR4nstSyVEXFi1cPDvAth3I9Rzzo01SURdAew7+3kfXmih/kVpm5z9QoDPLUehGmFvDs4wHuxhZ8XmS9UZRM/Qhxe4BnCTgdTKcrvRiO6o3A0Qk5GIw8xSaY6/w8nD03T
```
Comment (anandj91) [inline third_party/gpus/find_cuda_config.py.gz.base64:1]: what is this for?

File: third_party/gpus/generate_base64_for_find_cuda_config.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import zlib
+import base64
+
+with open("find_cuda_config.py", "rb") as reader:
+    code = base64.b64encode(zlib.compress(reader.read()))
+
+with open("find_cuda_config.py.gz.base64", "wb") as writer:
+    writer.write(code)
```
Comment (anandj91) [inline third_party/gpus/generate_base64_for_find_cuda_config.py:8]: what is this file for?

**Review by anandj91 — COMMENTED** [review]
I need to review the cutlass integration part more carefully. But meanwhile you can address other comments I have left.

File: tensorflow/compiler/xla/service/gpu/gemm_thunk.cc (hunk 5 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/gpu/gemm_thunk.cc]: was used for debugging, miss to remove. Done.

File: tensorflow/core/kernels/BUILD (hunk 8 above)
Comment (serach24) [reply tensorflow/core/kernels/BUILD]: Done

File: tensorflow/core/ops/math_ops.cc (hunk 9 above)
Comment (serach24) [reply tensorflow/core/ops/math_ops.cc:210]: Done

File: third_party/gpus/generate_base64_for_find_cuda_config.py (hunk 13 above)
Comment (serach24) [reply third_party/gpus/generate_base64_for_find_cuda_config.py:8]: This file is a script used to change content of find_cuda_config.py.gz.base64 based on content in find_cuda_config.py. In some .bzl files, it seems not able to run the python code directly, but only load the python script as binary to run it.

File: tensorflow/compiler/xla/service/gpu/gemm_thunk.cc (hunk 6 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/gpu/gemm_thunk.cc:297]: Resolved in the chat

File: tensorflow/core/platform/BUILD (hunk 10 above)
Comment (serach24) [reply tensorflow/core/platform/BUILD:149]: yes, resolved in chat

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc (hunk 2 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Done

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc (hunk 3 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Done

File: tensorflow/core/framework/types.h (hunk 7 above)
Comment (serach24) [reply tensorflow/core/framework/types.h]: done

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc [hunk 14]
```diff
@@ -1024,12 +993,22 @@ StatusOr<llvm::Value*> ElementalIrEmitter::EmitCusBinaryOp(
           return EmitCusGe(lhs_value, rhs_value, b_);
       }
     }
-    case HloOpcode::kMaximum:
-      return EmitFloatMax(lhs_value, rhs_value);
+    case HloOpcode::kMaximum:{
+      lhs_value = EmitCusToF32(lhs_value, b_);
+      rhs_value = EmitCusToF32(rhs_value, b_);
+      auto value = EmitFloatMax(lhs_value, rhs_value);
+      return EmitF32ToCus(value, b_);
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Shouldn't these be implemented inside the cus struct?

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc [hunk 15]
```diff
@@ -1024,12 +993,22 @@ StatusOr<llvm::Value*> ElementalIrEmitter::EmitCusBinaryOp(
...
+      
     // case HloOpcode::kMinimum:
     //   return EmitFloatMin(lhs_value, rhs_value);
-    // case HloOpcode::kPower:
-    //   return EmitPow(op->shape().element_type(), lhs_value, rhs_value);
+    case HloOpcode::kPower:{
+      lhs_value = EmitCusToF32(lhs_value, b_);
+      rhs_value = EmitCusToF32(rhs_value, b_);
+      auto value = EmitPow(op->shape().element_type(), lhs_value, rhs_value);
+      return EmitF32ToCus(value.ValueOrDie(), b_);
```
Comment (anandj91) [inline tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Shouldn't these be implemented inside the cus struct?

**Review by anandj91 — APPROVED** [review]

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc (hunk 15 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Done

File: tensorflow/compiler/xla/service/elemental_ir_emitter.cc (hunk 14 above)
Comment (serach24) [reply tensorflow/compiler/xla/service/elemental_ir_emitter.cc]: Done
